README
NAME
    Algorithm::Permute - Handy and fast permutation with object
    oriented interface

SYNOPSIS
      use Algorithm::Permute;

      my $p = new Algorithm::Permute(['a'..'d']);
      while (@res = $p->next) {
        print join(", ", @res), "\n";
      }

      my @array = (1..9);
      Algorithm::Permute::permute { print "@array\n" } @array;

DESCRIPTION
    This handy module makes performing permutation in Perl easy and
    fast, although perhaps its algorithm is not the fastest on the
    earth. Currently it only supports permutation n of n objects.

    No exported functions. This version is not backward compatible
    with the previous one, version 0.01. The old interface is no
    longer supported.

METHODS
    new [@list]
        Returns a permutor object for the given items.

    next
        Returns a list of the items in the next permutation. The
        order of the resulting permutation is the same as of the
        previous version of `Algorithm::Permute'.

    peek
        Returns the list of items which will be returned by next(),
        but doesn't advance the sequence. Could be useful if you
        wished to skip over just a few unwanted permutations.

    reset
        Resets the iterator to the start. May be used at any time,
        whether the entire set has been produced or not. Has no
        useful return value.

CALLBACK STYLE INTERFACE
    Starting with version 0.03, there is a function - not exported
    by default - which supports a callback style interface:

    permute BLOCK ARRAY
        A block of code is passed, which will be executed for each
        permutation. The array will be changed in place, and then
        changed back again before `permute' returns. During the
        execution of the callback, the array is read-only and you'll
        get an error if you try to change its length. (You *can*
        change its elements, but the consequences are liable to
        confuse you and may change in future versions.)

        You have to pass an array, it can't just be a list. It does
        work with special arrays and tied arrays, though unless
        you're doing something particularly abstruse you'd be better
        off copying the elements into a normal array first. Example:

         my @array = (1..9);
         permute { print "@array\n" } @array;

        The code is run inside a pseudo block, rather than as a
        normal subroutine. That means you can't use `return', and
        you can't jump out of it using `goto' and so on. Also,
        `caller' won't tell you anything helpful from inside the
        callback. Such is the price of speed.

        The order in which the permutations are generated is not
        guaranteed, so don't rely on it.

        The low-level hack behind this function makes it currently
        the fastest way of doing permutation among others.

COMPARISON
    I've collected some Perl routines and modules which implement
    permutation, and do some simple benchmark. The whole result is
    the following.

    Permutation of eight scalars:

     Abigail's:  9 wallclock secs ( 8.07 usr +  0.30 sys =  8.37 CPU)
    Algorithm::Permute:  5 wallclock secs ( 5.72 usr +  0.00 sys =  5.72 CPU)
    Algorithm::Permute qw(permute):  2 wallclock secs ( 1.65 usr +  0.00 sys =  1.65 CPU)
    List::Permutor: 27 wallclock secs (26.73 usr +  0.01 sys = 26.74 CPU)
         MJD's: 32 wallclock secs (32.55 usr +  0.02 sys = 32.57 CPU)
      perlfaq4: 36 wallclock secs (35.27 usr +  0.02 sys = 35.29 CPU)

    Permutation of nine scalars (the Abigail's routine is commented
    out, because it stores all of the result in memory, swallows all
    of my machine's memory):

     Algorithm::Permute: 43 wallclock secs (42.93 usr +  0.04 sys = 42.97 CPU)
     Algorithm::Permute qw(permute): 15 wallclock secs (14.82 usr +  0.00 sys = 14.82 CPU)
     List::Permutor: 227 wallclock secs (226.46 usr +  0.22 sys = 226.68 CPU)
         MJD's: 307 wallclock secs (306.69 usr +  0.43 sys = 307.12 CPU)
      perlfaq4: 272 wallclock secs (271.93 usr +  0.33 sys = 272.26 CPU)

    The benchmark script is included in the bench directory. I
    understand that speed is not everything. So here is the list of
    URLs of the alternatives, in case you hate this module.

    *   Mark Jason Dominus' technique is discussed in chapter 4 Perl
        Cookbook, so you can get it from O'Reilly:
        ftp://ftp.oreilly.com/published/oreilly/perl/cookbook

    *   Abigail's: http://www.foad.org/~abigail/Perl

    *   List::Permutor: http://www.cpan.org/modules/by-module/List

    *   The classic way, usually used by Lisp hackers: perldoc perlfaq4

HISTORY
    *   September 5, 2001 - version 0.03. A callback style interface,
        which is very fast, is added.

    *   September 2, 2000 - version 0.02. Major interface changes, now
        using object oriented interface similar to
        `List::Permutor''s. More efficient memory usage. Internal
        tweaking gives speed improvement - the list elements are no
        longer swapped, but only the indexes.

    *   October 3, 1999 - Alpha release, version 0.01

AUTHOR
    Edwin Pratomo, *ed.pratomo@computer.org*. The object oriented
    interface is taken from Tom Phoenix's `List::Permutor'. Robin
    Houston <robin@kitsite.com> invented and contributed the
    callback style interface.

ACKNOWLEDGEMENT
    Yustina Sri Suharini - my fiance, for providing the permutation
    problem to me.

SEE ALSO
    * Data Structures, Algorithms, and Program Style Using C - 
    Korsh and Garrett
    * Algorithms from P to NP, Vol. I - Moret and Shapiro


Changes
Revision history for Perl extension Algorithm::Permute.

0.11  Jan 4 2008
    - Added alternative implementation of the algorithm (enable it using
      -DUSE_LINKEDLIST)
    - Added comprehensive benchmarks script, thus replacing the old 
      stuffs inside bench directory

0.10  Jan 2 2008
    - Fixed memory error when cleaning up combination struct on Win32
    - Fixed compile error when using MSVC6

0.09  Jan 2 2008
    - Added test for die() inside permutation block
    - Added memory leak tests for A::P destructor
    - New feature: permutation r of n objects. Combination
      is implemented using coolex algorithm
    - Updated docs to reflect this feature
    - Fixed memory leak in A::P destructor

0.08  Dec 27 2007
    - Make it pass test #23 (srezic reported crashes on 
      amd64-freebsd-6.2). Thanks to Yoga Eko P for free login to 
      freebsd machine.

0.07  Dec 26 2007
    - Make it work with perl-5.10.0 (RT ticket #27022)

0.06  May 28 2003
    - Fixed a memory leak in permute()
    - Fixed bug #2596 : Perl context stack manip done in safer way in permute()
    - array passed to permute() not marked as R/O. see test #23 for reason
    - test.pl extended with eval and mem leak tests

0.05  May 16 2003
	- Added -DPERL_EXT to legalise use of perl-5.9.0's cxinc
	- Uses PAD_SET_CUR macro when defined, this fixes problem with invalid pad in 5.9.0

0.04  Mar 6 2002
	- Fixed cross-compiler compatibility problem in the callback hack, by Robin

0.03  Sep 5 2001
	- Added callback interface by Robin Houston
	- Fixed bogus benchmark

0.02  Sep 2 2000
	- Major interface changes. Now object-oriented

0.01  Fri Sep  1 18:05:17 2000
	- original version; created by h2xs 1.19



test.pl
# Before `make install' is performed this script should be runnable with
# `make test'. After `make install' it should work as `perl test.pl'

######################### We start with some black magic to print on failure.

# Change 1..1 below to 1..last_test_to_print .
# (It may become useful if the test is moved to ./t subdirectory.)

BEGIN { $| = 1; print "1..32\n"; }
END {print "not ok 1\n" unless $loaded;}
@correct = ("3 2 1", "2 3 1", "2 1 3", "3 1 2", "1 3 2", "1 2 3");

use Algorithm::Permute qw(permute);
$loaded = 1;
print "ok 1\n";

######################### End of black magic.

# Insert your test code below (better if it prints "ok 13"
# (correspondingly "not ok 13") depending on the success of chunk 13
# of the test code):

$perm = Algorithm::Permute->new([1..3]);
print ( $perm ? "ok 2\n" : "not ok 2\n");

# peek..
@peek = $perm->peek;
print "# @peek.\nnot " unless "@peek" eq $correct[0];
print "ok 3\n";

# next..
while (@res = $perm->next) {
    print "# @res.\nnot " unless "@res" eq $correct[$cnt++];
    print ("ok ". ($cnt + 3) . "\n");
}

# reset..
$cnt = 0;
$perm->reset;
while (@res = $perm->next) {
    print "# @res.\nnot " unless "@res" eq $correct[$cnt++];
    print "ok ". ($cnt + 9) . "\n";
}

print $cnt == 6 ? "ok 16\n" : "not ok 16\n";

# Tests for the callback interface by Robin Houston <robin@kitsite.com>

my @array = (1..9);
my $i = 0;
permute { ++$i } @array;

print ($i == 9*8*7*6*5*4*3*2*1 ? "ok 17\n" : "not ok 17\n");
print ($array[0] == 1 ? "ok 18\n" : "not ok 18\n");

@array = ();
$i = 0;
permute { ++$i } @array;
print ($i == 0 ? "ok 19\n" : "not ok 19\n");

@array = ('A'..'E');
my @foo;
permute { @foo = @array; } @array;

my $ok = ( join("", @foo) eq join("", reverse @array) );
print ($ok ? "ok 20\n" : "not ok 20\n");

tie @array, 'TieTest';
permute { $_ = "@array" } @array;
print (TieTest->c() == 600 ? "ok 21\n" : "not ok 21\t# ".TieTest->c()."\n");

untie @array;

##########################################
# test eval block outside of permute block
{
    my @array = (1..2);
    $i = 0;
    eval {
        permute {
            die if (++$i > 1 )
        } @array;
    };
    print "ok 22\n";
    eval { @array = (1..2); };                                      # try to change the array after die()
    print $@ ? 'not ' : '', "ok 23\n";
}

######################################
# test eval block inside permute block
@array = (qw/a r s e/);
$i = 0;
permute {eval {goto foo}; ++$i } @array;
if ($@ =~ /^Can't "goto" out/) {
    print "ok 24\n";
} else {
    foo: print "not ok 24\t# $@\n";
}
print ($i == 24 ? "ok 25\n" : "not ok 25\n");

{
    # test r of n permutation
    my %expected = map { $_ => 1 } qw/2_1 1_2 3_2 2_3 3_1 1_3/;
    my $p = Algorithm::Permute->new([1..3], 2);
    print ($p ? "ok 26\n" : "not ok 26\n");

    my $found;
    while (my @r = $p->next) {
        $found = delete $expected{join('_', @r)};
        break unless $found;
    }
    if (not $found or keys(%expected)) {
        print "not ok 27\n";
    } else {
        print "ok 27\n";
    }
}

######################
# test for memory leak

$^O !~ /linux/ || !$ENV{MEMORY_TEST} and do {
    for (28..32) { print "skipping $_: memory leak test\n" }
    exit 0;
};

# OO interface memory leak test
for ($i = 0;  $i < 10000;  $i++) {
    $perm->reset;
    while (@res = $perm->next) { }
    if ($i == 0) {
        $ok = check_mem(1);     # initialize
    }
    elsif ($i % 100  ==  99) {
        !$ok or $ok = check_mem();
    }
}
print $ok ? '' : 'not ', "ok 28\n";

for ($i = 0;  $i < 10000;  $i++) {
    @array = ('A'..'E');
    permute { } @array;

    if ($i == 0) {
        $ok = check_mem(1);     # initialize
    }
    elsif ($i % 100  ==  99) {
        !$ok or $ok = check_mem();
    }
}
print $ok ? '' : 'not ', "ok 29\n";

for ($i = 0;  $i < 10000;  $i++) {
    @array = ('A'..'E');
    eval { permute { die } @array };

    if ($i == 0) {
        $ok = check_mem(1);     # initialize
    }
    elsif ($i % 100  ==  99) {
        !$ok or $ok = check_mem();
    }
}
print $ok ? '' : 'not ', "ok 30\n";

{
    # test A::P destructor
    for ($i = 0;  $i < 10000;  $i++) {
        my $p = Algorithm::Permute->new([1..4]);
        while (@res = $p->next) { }
        if ($i == 0) {
            $ok = check_mem(1);     # initialize
        }
        elsif ($i % 100  ==  99) {
            !$ok or $ok = check_mem();
        }
    }
    print $ok ? '' : 'not ', "ok 31\n";
}

{
    # test A::P destructor, r of n permutation
    for ($i = 0;  $i < 10000;  $i++) {
        my $p = Algorithm::Permute->new([1..4], 3);
        while (@res = $p->next) { }
        if ($i == 0) {
            $ok = check_mem(1);     # initialize
        }
        elsif ($i % 100  ==  99) {
            !$ok or $ok = check_mem();
        }
    }
    print $ok ? '' : 'not ', "ok 32\n";
}

my $c;
package TieTest;
sub TIEARRAY  {bless []}
sub FETCHSIZE {5}
sub FETCH     { ++$c; $_[1]}
sub c         {$c}

package main;
sub check_mem {
    my $initialise = shift;
    # Log Memory Usage
    local $^W;
    my %mem;
    if (open(FH, "/proc/self/status")) {
        my $units;
        while (<FH>) {
            if (/^VmSize.*?(\d+)\W*(\w+)$/) {
                $mem{Total} = $1;
                $units = $2;
            }
            if (/^VmRSS:.*?(\d+)/) {
                $mem{Resident} = $1;
            }
        }
        close FH;

        print("# Mem Total: $mem{Total} $units, Resident: $mem{Resident} $units\n")
            if $ENV{MEMORY_TEST} > 1;

        if ($TOTALMEM != $mem{Total}) {
            warn("LEAK! : ", $mem{Total} - $TOTALMEM, " $units\n") unless $initialise;
            $TOTALMEM = $mem{Total};
            return $initialise ? 1 : 0;
        }

        return 1;
    }
}



coolex.c
/*
   Copyright (c) 2008  Edwin Pratomo

   You may distribute under the terms of either the GNU General Public
   License or the Artistic License, as specified in the Perl README file,
   with the exception that it cannot be placed on a CD-ROM or similar media
   for commercial distribution without the prior approval of the author.

*/

#include "coolex.h"

COMBINATION* init_combination(IV n, IV r, AV *av) {
    COMBINATION *c = NULL;
    int i;
    unsigned char *b = NULL;
    SV *aryref = newRV_inc((SV*) av);

    /* init bitstring */
    b = (unsigned char*)safecalloc(n, sizeof(unsigned char));
    if (b == NULL)
        return NULL;
    
    for (i = 0; i < r; i++)
        b[i] = 1;

    c = (COMBINATION*)safemalloc(sizeof(COMBINATION));
    if (c == NULL) {
        safefree(b);
        return NULL;
    }
    c->n = n;
    c->r = r;
    c->aryref = aryref;
    c->b = b;
    c->state = 0;
    return c;
}

void free_combination(COMBINATION *c) {
    safefree(c->b);
    SvREFCNT_dec(c->aryref);
    safefree(c);
}

/* coolex algorithm */
bool coolex(COMBINATION *c) {
    static int x = 1, y = 0;
    bool is_done = FALSE;
    
    switch (c->state) {
        case 0: /* state 0: initialized */
            c->state = 1; 
            break;
        case 1: /* state 1: first shift */
            c->b[c->r] =1;
            c->b[0] = 0;
            c->state = 2;
            break;
        default: /* subsequent shifts */
        {
            while (x < c->n - 1) {
                c->b[x++] = 0;
                c->b[y++] = 1;
                if (c->b[x] == 0) {
                    c->b[x] = 1, c->b[0] = 0;
                    if (y > 1) x = 1;
                    y = 0;
                }
                return is_done;
            }
            /* reset x and y */
            x = 1, y = 0;
            is_done = TRUE;
        }
    }    
    return is_done;
}

void coolex_visit(COMBINATION *c, SV **p_items) {
    int i, r = 0;
    SV **p, **svp;
    AV *av = (AV*)SvRV(c->aryref);
    for (i = 0, p = p_items; i < c->n; i++) {
        if (c->b[i]) {    /* the bitstring matters */
            r++;
            /* tell GC to take care of this */
            if (SvOK(*p)) {
                SvREFCNT_dec(*p);
            }
            svp = av_fetch(av, i, FALSE);
            *p = (svp) ? SvREFCNT_inc(*svp) : &PL_sv_undef;
            p++;
        }
    } 
    assert(r == c->r);
}



coolex.h
/*
   Copyright (c) 2008  Edwin Pratomo

   You may distribute under the terms of either the GNU General Public
   License or the Artistic License, as specified in the Perl README file,
   with the exception that it cannot be placed on a CD-ROM or similar media
   for commercial distribution without the prior approval of the author.

*/

#ifndef COOLEX_H
#define COOLEX_H
#include "EXTERN.h"
#include "perl.h"
#include "XSUB.h"

typedef struct {
	IV 	            n;
	IV              r;
	SV              *aryref;
    unsigned char   *b;             /* bitstring: array of bytes */
	int             state;          /* state 0 / 1 / 2 */
} COMBINATION;

COMBINATION* init_combination(IV n, IV r, AV *av);
void free_combination(COMBINATION *c);
/* coolex pseudo-coroutine */
bool coolex(COMBINATION *c);
void coolex_visit(COMBINATION *c, SV **p_items);

#endif


coollex.c
/*
   Copyright (c) 2008  Edwin Pratomo

   You may distribute under the terms of either the GNU General Public
   License or the Artistic License, as specified in the Perl README file,
   with the exception that it cannot be placed on a CD-ROM or similar media
   for commercial distribution without the prior approval of the author.

*/

#include "coollex.h"

COMBINATION* init_combination(IV n, IV r, AV *av) {
    COMBINATION *c = NULL;
    int i;
    unsigned char *b = NULL;
    SV *aryref = newRV_inc((SV*) av);

    /* init bitstring */
    b = (unsigned char*)safecalloc(n, sizeof(unsigned char));
    if (b == NULL)
        return NULL;
    
    for (i = 0; i < r; i++)
        b[i] = 1;

    c = (COMBINATION*)safemalloc(sizeof(COMBINATION));
    if (c == NULL) {
        safefree(b);
        return NULL;
    }
    c->n = n;
    c->r = r;
    c->aryref = aryref;
    c->b = b;
    c->state = 0;
    return c;
}

void free_combination(COMBINATION *c) {
    safefree(c->b);
    SvREFCNT_dec(c->aryref);
    safefree(c);
}

/* coollex algorithm */
bool coollex(COMBINATION *c) {
    static int x = 1, y = 0;
    bool is_done = FALSE;
    
    switch (c->state) {
        case 0: /* state 0: initialized */
            c->state = 1; 
            break;
        case 1: /* state 1: first shift */
            c->b[c->r] =1;
            c->b[0] = 0;
            c->state = 2;
            break;
        default: /* subsequent shifts */
        {
            while (x < c->n - 1) {
                c->b[x++] = 0;
                c->b[y++] = 1;
                if (c->b[x] == 0) {
                    c->b[x] = 1, c->b[0] = 0;
                    if (y > 1) x = 1;
                    y = 0;
                }
                return is_done;
            }
            /* reset x and y */
            x = 1, y = 0;
            is_done = TRUE;
        }
    }    
    return is_done;
}

void coollex_visit(COMBINATION *c, SV **p_items) {
    int i, r = 0;
    SV **p, **svp;
    AV *av = (AV*)SvRV(c->aryref);
    for (i = 0, p = p_items; i < c->n; i++) {
        if (c->b[i]) {    /* the bitstring matters */
            r++;
            /* tell GC to take care of this */
            if (SvOK(*p)) {
                SvREFCNT_dec(*p);
            }
            svp = av_fetch(av, i, FALSE);
            *p = (svp) ? SvREFCNT_inc(*svp) : &PL_sv_undef;
            p++;
        }
    } 
    assert(r == c->r);
}



coollex.h
/*
   Copyright (c) 2008  Edwin Pratomo

   You may distribute under the terms of either the GNU General Public
   License or the Artistic License, as specified in the Perl README file,
   with the exception that it cannot be placed on a CD-ROM or similar media
   for commercial distribution without the prior approval of the author.

*/

#ifndef COOLEX_H
#define COOLEX_H
#include "EXTERN.h"
#include "perl.h"
#include "XSUB.h"

typedef struct {
	IV 	            n;
	IV              r;
	SV              *aryref;
    unsigned char   *b;             /* bitstring: array of bytes */
	int             state;          /* state 0 / 1 / 2 */
} COMBINATION;

COMBINATION* init_combination(IV n, IV r, AV *av);
void free_combination(COMBINATION *c);
/* coollex pseudo-coroutine */
bool coollex(COMBINATION *c);
void coollex_visit(COMBINATION *c, SV **p_items);

#endif


Permute.pm
#   Permute.pm
#
#   Copyright (c) 1999 - 2008 Edwin Pratomo
#
#   You may distribute under the terms of either the GNU General Public
#   License or the Artistic License, as specified in the Perl README file,
#   with the exception that it cannot be placed on a CD-ROM or similar media
#   for commercial distribution without the prior approval of the author.

package Algorithm::Permute;

use strict;
use vars qw($VERSION @ISA @EXPORT @EXPORT_OK);

require Exporter;
require DynaLoader;
require AutoLoader;

@ISA = qw(Exporter DynaLoader);
@EXPORT_OK = qw(permute);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

$VERSION = '0.11';

bootstrap Algorithm::Permute $VERSION;

# Preloaded methods go here.

# Autoload methods go after =cut, and are processed by the autosplit program.

1;
__END__

=head1 NAME

Algorithm::Permute - Handy and fast permutation with object oriented interface

=head1 SYNOPSIS

  use Algorithm::Permute;

  # default is to create n of n objects permutation generator
  my $p = new Algorithm::Permute(['a'..'d']);

  # but also you can create r of n objects permutation generator, where r <= n
  my $p = new Algorithm::Permute([1..4], 3);

  while (@res = $p->next) {
    print join(", ", @res), "\n";
  }

  # and this one is the speed demon:
  my @array = (1..9);
  Algorithm::Permute::permute { print "@array\n" } @array;

=head1 DESCRIPTION

This handy module makes performing permutation in Perl easy and fast, 
although perhaps its algorithm is not the fastest on the earth. 
It supports permutation r of n objects where 0 < r <= n. 

=head1 METHODS

=over 4

=item new [@list]

Returns a permutor object for the given items. 

=item next

Returns a list of the items in the next permutation. 
The order of the resulting permutation is the same as of the previous version 
of C<Algorithm::Permute>.

=item peek

Returns the list of items which B<will be returned> by next(), but
B<doesn't advance the sequence>. Could be useful if you wished to skip
over just a few unwanted permutations.

=item reset

Resets the iterator to the start. May be used at any time, whether the
entire set has been produced or not. Has no useful return value.

=back

=head1 CALLBACK STYLE INTERFACE

Starting with version 0.03, there is a function - not exported by
default - which supports a callback style interface:

=over 4

=item permute BLOCK ARRAY

A block of code is passed, which will be executed for each permutation. The array will be changed in place,
and then changed back again before C<permute> returns. During the execution of the callback,
the array is read-only and you'll get an error if you try to change its length. (You I<can>
change its elements, but the consequences are liable to confuse you and may change in future
versions.)

You have to pass an array, it can't just be a list. It B<does> work with special arrays
and tied arrays, though unless you're doing something particularly abstruse you'd be
better off copying the elements into a normal array first. Example:

 my @array = (1..9);
 permute { print "@array\n" } @array;

The code is run inside a pseudo block, rather than as a normal subroutine. That means
you can't use C<return>, and you can't jump out of it using C<goto> and so on. Also,
C<caller> won't tell you anything helpful from inside the callback. Such is the price
of speed.

The order in which the permutations are generated is not guaranteed, so don't rely
on it. 

The low-level hack behind this function makes it currently the fastest way of doing
permutation among others. 

=back

=head1 COMPARISON

I've collected some Perl routines and modules which implement permutation,
and do some simple benchmark. The whole result is the following.

Permutation of B<eight> scalars:

  Abigail's                     :  9 wallclock secs ( 8.07 usr +  0.30 sys =  8.37 CPU)
  Algorithm::Permute            :  5 wallclock secs ( 5.72 usr +  0.00 sys =  5.72 CPU)
  Algorithm::Permute qw(permute):  2 wallclock secs ( 1.65 usr +  0.00 sys =  1.65 CPU)
  List::Permutor                : 27 wallclock secs (26.73 usr +  0.01 sys = 26.74 CPU)
  Memoization                   : 32 wallclock secs (32.55 usr +  0.02 sys = 32.57 CPU)
  perlfaq4                      : 36 wallclock secs (35.27 usr +  0.02 sys = 35.29 CPU)

Permutation of B<nine> scalars (the Abigail's routine is commented out, because
it stores all of the result in memory, swallows all of my machine's memory):

  Algorithm::Permute            :  43 wallclock secs ( 42.93 usr +  0.04 sys = 42.97 CPU)
  Algorithm::Permute qw(permute):  15 wallclock secs ( 14.82 usr +  0.00 sys = 14.82 CPU)
  List::Permutor                : 227 wallclock secs (226.46 usr +  0.22 sys = 226.68 CPU)
  Memoization                   : 307 wallclock secs (306.69 usr +  0.43 sys = 307.12 CPU)
  perlfaq4                      : 272 wallclock secs (271.93 usr +  0.33 sys = 272.26 CPU)

The benchmark script is included in the bench directory. I understand that 
speed is not everything. So here is the list of URLs of the alternatives, in 
case you hate this module.

=over 4

=item * 

Memoization is discussed in chapter 4 Perl Cookbook, so you can get it from
O'Reilly: ftp://ftp.oreilly.com/published/oreilly/perl/cookbook

=item *

Abigail's: http://www.foad.org/~abigail/Perl

=item *

List::Permutor: http://www.cpan.org/modules/by-module/List

=item *

The classic way, usually used by Lisp hackers: perldoc perlfaq4

=back

=head1 AUTHOR

Edwin Pratomo, I<edpratomo@cpan.org>. 

The object oriented interface is taken from Tom Phoenix's C<List::Permutor>.
Robin Houston <robin@kitsite.com> invented and contributed the callback
style interface.

=head1 ACKNOWLEDGEMENT

Yustina Sri Suharini - my ex-fiance-now-wife, for providing the permutation
problem to me.

=head1 SEE ALSO

=over 2

=item * B<Data Structures, Algorithms, and Program Style Using C> - 
Korsh and Garrett

=item * B<Algorithms from P to NP, Vol. I> - Moret and Shapiro

=back

=cut
